### PR TITLE
fix hosting.config reload

### LINKED ISSUE #7220: hosting.config requires restart, but docs say reloadable
Docs say that hosting.config changes are reloadable:
https://docs.trafficserver.apache.org/en/8.0.x/admin-guide/files/hosting.config.en.html
"After you modify hosting.config, navigate to the Traffic Server bin directory and run traffic_ctl config reload to apply your changes."
This line has been in the docs since the initial release.

In testing, we observe that changes to hosting.config only take effect after restart. The best way I've found to see this:
- make a host with 2 volumes. In hosting.config have
```hostname=* volume=1```
- after starting up, fetch objects. ```traffic_ctl metric match volume | grep bytes_used``` should show ```proxy.process.cache.volume_1.bytes_used``` increasing.
- edit hosting.config, adding the origin (for example ```hostname=origin.example.org volume=2```
- reload trafficserver
- fetch more objects. You'd expect ```proxy.process.cache.volume_2.bytes_used``` to now increase, however, volume_1 continues to increase instead.
- Upon restart and more fetches, ```proxy.process.cache.volume_2.bytes_used``` begins increasing.

Our primary use case for the hosting.config is to assign "properties" that either have low max-age or useful lifetime (for instance a live sports stat object may be constantly updated; a live video segment is only fetched in a small time window). Being able to dynamically add such use cases would be useful.

### PR DESCRIPTION
**Pull request by ezelkow1**
When hosting.config reloads, currently the file is properly loaded and placed in a new hosting table. However when ats goes to place it in use in the current cache structure it's using an ink_atomic_swap(new_table, old_table). This function does not actually swap, it only does a set, so it effectively overwrites the new table with the old table, and then continues, never updating. This is why reloading with a new hosting.config never sees an update in use.

This PR changes that to get rid of the ink_atomic use and just use a std_atomic for the table, replaces the host table ptr, and deletes the old one

closes #7220

### CODE REVIEW

**Comment by ezelkow1** [comment]
[approve ci centos]

**Comment by ezelkow1** [comment]
[approve ci rocky]

**Comment by ezelkow1** [comment]
suggestion from @SolidWallOfCode to take a look at the ipallow logic and see if we can do similar here

**Comment by ezelkow1** [comment]
Finally getting back to this, using the ipallow logic would take a very large rewrite from what I can tell. IMHO we should just do the simplest fix of swapping the values to the 'atomic_swap', the way it was meant to be used in the first place, to at least have this working. Then when we get around to a full hosting rewrite can look at doing it better then

**Review by SolidWallOfCode — DISMISSED** [review]

**Comment by SolidWallOfCode** [comment]
What we should really do is change it to a `std::shared_ptr` with a read/write mutex. This fix retains the arbitrary timeout which could lead to issues.

**Comment by ezelkow1** [comment]
[approve ci]

File: iocore/cache/P_CacheHosting.h [hunk 1]
```diff
@@ -163,8 +163,9 @@ struct CacheHostTableConfig : public Continuation {
   {
     (void)e;
     (void)event;
-    CacheHostTable *t   = new CacheHostTable((*ppt)->cache, (*ppt)->type);
-    CacheHostTable *old = (CacheHostTable *)ink_atomic_swap(&t, *ppt);
+    CacheHostTable *t = new CacheHostTable((*ppt)->cache, (*ppt)->type);
+    auto old          = *ppt;
+    (*ppt)            = t;
```
Comment (rob05c) [inline iocore/cache/P_CacheHosting.h:168]: I think this needs to be a `std::atomic<CacheHostTable *>* ppt` or possibly `std::atomic<std::atomic<CacheHostTable *>*> ppt;`.

With the outer pointer-pointer being atomic, but not the inner, I believe in `auto old = *ppt; the dereference `*ppt` will be atomic, but not the assignment to `old`. Likewise, in `(*ppt) = t;` I believe the `*ppt` dereference will be atomic, but when it takes `t` and assigns it to the value-pointer-to-by `ppt`, I don't think that will be atomic.

File: iocore/cache/P_CacheHosting.h [hunk 2]
```diff
@@ -152,7 +152,7 @@ class CacheHostTable
 struct CacheHostTableConfig;
 typedef int (CacheHostTableConfig::*CacheHostTabHandler)(int, void *);
 struct CacheHostTableConfig : public Continuation {
-  CacheHostTable **ppt;
+  std::atomic<CacheHostTable **> ppt;
```
Comment (rob05c) [inline iocore/cache/P_CacheHosting.h:155]: I noticed this isn't used by anything external. It'd be nice to make it `private`, to enforce that, and make it easier for readers to know that

**Comment by randall** [comment]
Is this pr still needed now that https://github.com/apache/trafficserver/pull/9046 has been merged?

**Comment by ezelkow1** [comment]
> Is this pr still needed now that #9046 has been merged?

nope, closing

**Pull request closed without merging**